### ISSUE TITLE
How to parse an InChi string?

### ISSUE THREAD

**Issue by tomas-pluskal** [cdk/cdk#475]
Hi,

I have a String with an InChi, e.g.

````
String inchi = "InChI=1S/C17H23NO2/c1-13-17(2,3)14-9-6-7-10-15(14)18(13)12-8-4-5-11-16(19)20/h6-7,9-10H,1,4-5,8,11-12H2,2-3H3,(H,19,20)";
````
and I would like to parse this InChi into an IAtomContainer. I thought this would be a trivial operation using CDK, but somehow I cannot find a way to do it...?

After some desperate attempts I came up with this, which seems overly complicated and also drops all hydrogens from the molecule:
````
        StringReader stringReader = new StringReader(inchi);
        INChIPlainTextReader inchiReader =
            new INChIPlainTextReader(stringReader);        
        IChemFile cFile = (IChemFile) inchiReader.read(new ChemFile());
        List<IAtomContainer> containers = ChemFileManipulator.getAllAtomContainers(cFile);
        IAtomContainer molecule = containers.get(0);
````

Is there a better way, something similar to ```SmilesParser. parseSmiles()```?

Tomas

**Comment by tomas-pluskal** [comment]
Thanks!

> Please remember you should not (and can not in all cases) read InChIs.

Why?

**Comment by johnmay** [comment]
Because​ it's an identifier and the conversion is lossy. From the direct of
the InChI project: "InChI is not a replacement for any existing internal
structure representations. (We do not start religious wars. ) InChI is in
ADDITION to what one uses internally. Its value to student or scientist is
in FINDING and LINKING information"

http://www.hellers.com/steve/pub-talks/toronto-7-14.pdf

You're probably still going to to read them so I won't stop you but really
it's not meant for that.

**Comment by tomas-pluskal** [comment]
Thanks, John. I don't really have any alternative to reading InChis, since I was given a database of molecules as a table of InChi strings.

**Comment by tomas-pluskal** [comment]
But in practice, does it matter? I mean, as long as the molecules don't contain any structural features that cannot be represented in InChi, I don't see any problem. I also don't see any better alternative (SMILES?).

**Comment by egonw** [comment]
@tomas-pluskal, did you consider using the InChI to just look up the SMILES in PubChem/ChemSpider/... ?

BTW #1, using the InChI as identifier in databases is indeed something else than using it as structural representation... many databases only need the identifier, because they don't understand the chemistry anyway (sad truth...) 

BTW #2, we do have an InChIParser (or used to), but it should be clear is what you get back is a rough structure, but with undefined info... now, that never worried me, as we have multiple formats with limited information, and even more with implicit information by format standards (where you can only hope the one who wrote in that format was aware of that...)...

BTW #3, I do like to be able to parse the InChI too, to extract the layers... so that I can compare InChIs at a layer-by-layer level, to identify equivalence... tempted to write one this summer...

**Comment by tomas-pluskal** [comment]
So, I have been looking at https://www.inchi-trust.org/technical-faq/, and it is still not quite clear to me what are the limitations of InChi. They say there is a "Standard InChI" which provides a standardized representation of the molecule. Is there less information in "Standard InChI" than in SMILES?

**Comment by johnmay** [comment]
Yes and no but the point is the structure is changed to be useful for computer to look at and not a human. For example peptide bonds come out wrong and all metals disconnection. However, If it works for you, great.
